### PR TITLE
Fix #6446: Fix lock contention ANR in AudioFragment and AudioPlayerController

### LINKED ISSUE #6446: [BUG]: Input Lock Contention Issues in AudioFragment
### Describe the bug

The main thread was blocked trying to acquire a lock that was held by another thread.

**The main thread is blocked here. It is waiting, trying to acquire lock 0x0c54a4c1 (android.media.MediaHTTPConnection), held by thread 58.** 

```php
at android.media.IMediaHTTPConnection$Stub.onTransact (IMediaHTTPConnection.java:137)
  at android.os.Binder.execTransactInternal (Binder.java:1344)
  at android.os.Binder.execTransact (Binder.java:1275)
  at android.media.MediaPlayer._reset (Native method)
  at android.media.MediaPlayer.reset (MediaPlayer.java:2275)
  at org.oppia.android.domain.audio.AudioPlayerController.changeDataSource (AudioPlayerController.java:123)
  at org.oppia.android.app.player.audio.AudioViewModel.loadAudio (AudioViewModel.java:183)
  at org.oppia.android.app.player.audio.AudioViewModel.maybeLoadAudio (AudioViewModel.java:123)
  at org.oppia.android.app.player.audio.AudioViewModel.setStateAndExplorationId (AudioViewModel.java:73)
  at org.oppia.android.app.player.audio.AudioFragmentPresenter.setStateAndExplorationId (AudioFragmentPresenter.java:226)
  at org.oppia.android.app.player.audio.AudioFragment.setStateAndExplorationId (AudioFragment.java:91)
  at org.oppia.android.app.player.state.StatePlayerRecyclerViewAssembler.compute (StatePlayerRecyclerViewAssembler.java:398)
  at org.oppia.android.app.player.state.StateFragmentPresenter.processEphemeralState (StateFragmentPresenter.java:361)
  at org.oppia.android.app.player.state.StateFragmentPresenter.processEphemeralStateResult (StateFragmentPresenter.java:337)
  at org.oppia.android.app.player.state.StateFragmentPresenter.access$processEphemeralStateResult (StateFragmentPresenter.java:73)
  at org.oppia.android.app.player.state.StateFragmentPresenter$subscribeToCurrentState$1.onChanged (StateFragmentPresenter.java:327)
  at org.oppia.android.app.player.state.StateFragmentPresenter$subscribeToCurrentState$1.onChanged (StateFragmentPresenter.java:324)
  at androidx.lifecycle.LiveData.considerNotify (LiveData.java:133)
```

**Thread 58 "binder:17156_4" tid=58 Native**
```php
at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress (RouteSelector.java:178)
  at com.android.okhttp.internal.http.RouteSelector.nextProxy (RouteSelector.java:144)
  at com.android.okhttp.internal.http.RouteSelector.next (RouteSelector.java:86)
  at com.android.okhttp.internal.http.StreamAllocation.findConnection (StreamAllocation.java:176)
  at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection (StreamAllocation.java:128)
  at com.android.okhttp.internal.http.StreamAllocation.newStream (StreamAllocation.java:97)
  at com.android.okhttp.internal.http.HttpEngine.connect (HttpEngine.java:289)
  at com.android.okhttp.internal.http.HttpEngine.sendRequest (HttpEngine.java:232)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute (HttpURLConnectionImpl.java:465)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse (HttpURLConnectionImpl.java:411)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode (HttpURLConnectionImpl.java:542)
  at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode (DelegatingHttpsURLConnection.java:106)
  at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode (HttpsURLConnectionImpl.java:27)
  at android.media.MediaHTTPConnection.seekTo (MediaHTTPConnection.java:285)
```

**Lock 0x0c54a4c1 (android.media.MediaHTTPConnection), was acquired by this thread here. This is causing an ANR, as the main thread is blocked trying to acquire the same lock.**

```php
at android.media.MediaHTTPConnection.getMIMEType (MediaHTTPConnection.java:460)
  at android.media.IMediaHTTPConnection$Stub.onTransact (IMediaHTTPConnection.java:162)
  at android.os.Binder.execTransactInternal (Binder.java:1344)
  at android.os.Binder.execTransact (Binder.java:1275)
```

**The main thread is blocked here. It is waiting, trying to acquire lock 0x0b97fc9e (android.media.MediaHTTPConnection), held by thread 5.** 

```php
at android.media.IMediaHTTPConnection$Stub.onTransact (IMediaHTTPConnection.java:132)
  at android.os.Binder.execTransactInternal (Binder.java:1159)
  at android.os.Binder.execTransact (Binder.java:1123)
  at android.media.MediaPlayer._reset (Native method)
  at android.media.MediaPlayer.reset (MediaPlayer.java:2146)
  at org.oppia.android.domain.audio.AudioPlayerController.changeDataSource (AudioPlayerController.java:123)
  at org.oppia.android.app.player.audio.AudioViewModel.loadAudio (AudioViewModel.java:183)
  at org.oppia.android.app.player.audio.AudioViewModel.maybeLoadAudio (AudioViewModel.java:123)
  at org.oppia.android.app.player.audio.AudioViewModel.setStateAndExplorationId (AudioViewModel.java:73)
  at org.oppia.android.app.player.audio.AudioFragmentPresenter.setStateAndExplorationId (AudioFragmentPresenter.java:226)
  at org.oppia.android.app.player.audio.AudioFragment.setStateAndExplorationId (AudioFragment.java:91)
  at org.oppia.android.app.player.state.StatePlayerRecyclerViewAssembler.compute (StatePlayerRecyclerViewAssembler.java:398)
  at org.oppia.android.app.player.state.StateFragmentPresenter.processEphemeralState (StateFragmentPresenter.java:361)
  at org.oppia.android.app.player.state.StateFragmentPresenter.processEphemeralStateResult (StateFragmentPresenter.java:337)
  at org.oppia.android.app.player.state.StateFragmentPresenter.access$processEphemeralStateResult (StateFragmentPresenter.java:73)
  at org.oppia.android.app.player.state.StateFragmentPresenter$subscribeToCurrentState$1.onChanged (StateFragmentPresenter.java:327)
  at org.oppia.android.app.player.state.StateFragmentPresenter$subscribeToCurrentState$1.onChanged (StateFragmentPresenter.java:324)
  at androidx.lifecycle.LiveData.considerNotify (LiveData.java:133)
  at androidx.lifecycle.LiveData.dispatchingValue (LiveData.java:151)
  at androidx.lifecycle.LiveData.setValue (LiveData.java:309)
  at org.oppia.android.util.data.DataProviders$NotifiableAsyncLiveData.setValue (DataProviders.java:405)
  at org.oppia.android.util.data.DataProviders$NotifiableAsyncLiveData.setValue (DataProviders.java:354)
  at androidx.lifecycle.LiveData$1.run (LiveData.java:93)
  at android.os.Handler.handleCallback (Handler.java:938)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loop (Looper.java:223)
  at android.app.ActivityThread.main (ActivityThread.java:7945)
  at java.lang.reflect.Method.invoke (Native method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:603)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:947)
```

**Thread 5 "Binder:11261_4" tid=5 Native**

```php
at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress (RouteSelector.java:178)
  at com.android.okhttp.internal.http.RouteSelector.nextProxy (RouteSelector.java:144)
  at com.android.okhttp.internal.http.RouteSelector.next (RouteSelector.java:86)
  at com.android.okhttp.internal.http.StreamAllocation.findConnection (StreamAllocation.java:176)
  at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection (StreamAllocation.java:128)
  at com.android.okhttp.internal.http.StreamAllocation.newStream (StreamAllocation.java:97)
  at com.android.okhttp.internal.http.HttpEngine.connect (HttpEngine.java:289)
  at com.android.okhttp.internal.http.HttpEngine.sendRequest (HttpEngine.java:232)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute (HttpURLConnectionImpl.java:465)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse (HttpURLConnectionImpl.java:411)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode (HttpURLConnectionImpl.java:542)
  at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode (DelegatingHttpsURLConnection.java:106)
  at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode (HttpsURLConnectionImpl.java:30)
  at android.media.MediaHTTPConnection.seekTo (MediaHTTPConnection.java:285)
```

**Lock 0x0b97fc9e (android.media.MediaHTTPConnection), was acquired by this thread here. This is causing an ANR, as the main thread is blocked trying to acquire the same lock.** 

```php
  at android.media.MediaHTTPConnection.getMIMEType (MediaHTTPConnection.java:460)
  at android.media.IMediaHTTPConnection$Stub.onTransact (IMediaHTTPConnection.java:159)
  at android.os.Binder.execTransactInternal (Binder.java:1159)
  at android.os.Binder.execTransact (Binder.java:1123)
```

**The main thread is blocked here. It is waiting, trying to acquire lock 0x06de8cb8 (android.media.MediaHTTPConnection), held by thread 12.**

```php
t android.media.IMediaHTTPConnection$Stub.onTransact (IMediaHTTPConnection.java:134)
  at android.os.Binder.execTransactInternal (Binder.java:1226)
  at android.os.Binder.execTransact (Binder.java:1163)
  at android.media.MediaPlayer._release (Native method)
  at android.media.MediaPlayer.release (MediaPlayer.java:2218)
  at org.oppia.android.domain.audio.AudioPlayerController.releaseMediaPlayer (AudioPlayerController.java:260)
  at org.oppia.android.app.player.audio.AudioViewModel.handleRelease (AudioViewModel.java:242)
  at org.oppia.android.app.player.audio.AudioFragmentPresenter.handleOnDestroy (AudioFragmentPresenter.java:221)
  at org.oppia.android.app.player.audio.AudioFragment.onDestroy (AudioFragment.java:73)
  at androidx.fragment.app.Fragment.performDestroy (Fragment.java:3219)
  at androidx.fragment.app.FragmentStateManager.destroy (FragmentStateManager.java:774)
  at androidx.fragment.app.FragmentStateManager.moveToExpectedState (FragmentStateManager.java:350)
  at androidx.fragment.app.FragmentStore.moveToExpectedState (FragmentStore.java:112)
  at androidx.fragment.app.FragmentManager.moveToState (FragmentManager.java:1647)
  at androidx.fragment.app.FragmentManager.dispatchStateChange (FragmentManager.java:3128)
  at androidx.fragment.app.FragmentManager.dispatchDestroy (FragmentManager.java:3107)
  at androidx.fragment.app.Fragment.performDestroy (Fragment.java:3214)
  at androidx.fragment.app.FragmentStateManager.destroy (FragmentStateManager.java:774)
  at androidx.fragment.app.FragmentStateManager.moveToExpectedState (FragmentStateManager.java:350)
  at androidx.fragment.app.FragmentStore.moveToExpectedState (FragmentStore.java:112)
  at androidx.fragment.app.FragmentManager.moveToState (FragmentManager.java:1647)
  at androidx.fragment.app.FragmentManager.dispatchStateChange (FragmentManager.java:3128)
  at androidx.fragment.app.FragmentManager.dispatchDestroy (FragmentManager.java:3107)
  at androidx.fragment.app.Fragment.performDestroy (Fragment.java:3214)
  at androidx.fragment.app.FragmentStateManager.destroy (FragmentStateManager.java:774)
```

**Thread 12 "Binder:6921_5" tid=12 Native Holding lock**

```php
  at com.android.okhttp.internal.http.HttpEngine.connect (HttpEngine.java:289)
  at com.android.okhttp.internal.http.HttpEngine.sendRequest (HttpEngine.java:232)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute (HttpURLConnectionImpl.java:465)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse (HttpURLConnectionImpl.java:411)
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode (HttpURLConnectionImpl.java:542)
  at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode (DelegatingHttpsURLConnection.java:106)
  at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode (HttpsURLConnectionImpl.java:30)
  at android.media.MediaHTTPConnection.seekTo (MediaHTTPConnection.java:302)
```

**Lock 0x06de8cb8 (android.media.MediaHTTPConnection), was acquired by this thread here. This is causing an ANR, as the main thread is blocked trying to acquire the same lock.** 

```php
  at android.media.MediaHTTPConnection.readAt (MediaHTTPConnection.java:418)
  at android.media.MediaHTTPConnection.native_readAt (Native method)
  at android.media.MediaHTTPConnection.readAt (MediaHTTPConnection.java:407)
  at android.media.IMediaHTTPConnection$Stub.onTransact (IMediaHTTPConnection.java:145)
  at android.os.Binder.execTransactInternal (Binder.java:1226)
  at android.os.Binder.execTransact (Binder.java:1163)
```

### Steps To Reproduce

This seems to be happening when users try to play audio during a lesson.

### Expected Behavior

The audio should play without causing the app to crash.

### Screenshots/Videos

_No response_

### What device/emulator are you using?

Infinix-X6525D, OPPO OP56F5, Redmi gale, Redmi maltose, IPRO SAMBA65S

### Which Android version is your device/emulator running?

11, 12,13,14

### Which version of the Oppia Android app are you using?

0.18-rc04-beta-c495933008

### Additional Context

Reference: https://developer.android.com/topic/performance/anrs/find-unresponsive-thread#lock-contention)

This is a user perceieved ANR that has affected 10 users so far.

**Possible Cause**

The ANR is likely caused by `MediaPlayer.reset()` and `MediaPlayer.release()` being called synchronously on the main thread while an audio network operation is still in progress.

Both operations can wait for Android's `MediaHTTPConnection` lock. If the media networking thread is performing slow connection, DNS, or read work while holding that lock, the main thread can remain blocked long enough to trigger an ANR.

The traces show this occurring through two main paths: changing the lesson audio and leaving the lesson(i.e., destroying the audio fragment).

- Changing lesson audio: state updates reach `AudioViewModel.loadAudio()`, then `changeDataSource()` (line 117). Line 123 calls `mediaPlayer.reset()` directly. In both reset traces, a Binder thread holds the MediaHTTPConnection lock while getMIMEType() performs HTTP connection work.
- Leaving the lesson: AudioFragment.onDestroy() calls through to `releaseMediaPlayer()` (line 253) Line 260 calls `mediaPlayer.release()` directly. The third trace shows this waiting on the same kind of lock, held during `readAt()` network work.
- `AudioViewModel.setStateAndExplorationId()` triggers audio loading without first checking whether the requested source is already the currently loaded source. Repeated state updates could therefore cause unnecessary reloads and additional calls to reset() while an existing request is in progress.

**Suggested Fix**

Move MediaPlayer operations off the main thread.

A robust approach would be to give MediaPlayer a dedicated background thread with a Looper and serialize the player's lifecycle on that thread. This ensures that a slow reset() or release() waiting for Android's media networking layer does not freeze the UI.

The implementation should also avoid making the main thread wait on a lock held while MediaPlayer operations are running.

Additionally, `AudioViewModel` can avoid restarting audio loading when the requested source is already current.

### PR DESCRIPTION
**Pull request by Kishan8548**
## Explanation
Fixes #6446

Fixes input lock contention and ANR issues in `AudioFragment` and `AudioPlayerController`.

### Root Cause
1. Synchronous calls to `MediaPlayer.reset()` and `MediaPlayer.release()` were executed on the main UI thread. During network I/O, Android's internal `MediaHTTPConnection` binder thread holds internal locks, causing the main thread to block and trigger ANRs.
2. `AudioViewModel.loadAudio()` was re-triggered on every ephemeral state compute even if the requested voiceover audio URI was already loaded, causing redundant resets and re-loads.

### Solution
1. **AudioViewModel**:
   - Tracked `currentLoadedAudioUri` to short-circuit redundant `loadAudio()` invocations when the target URI is already active.
   - Cleared the tracked URI on reset, release, and abort.
2. **AudioPlayerController**:
   - Injected `@BackgroundDispatcher` to offload blocking `MediaPlayer` operations (`reset()`, `prepareDataSource()`, and `release()`) via Kotlin coroutines.
   - Used a suspending `Mutex` to serialize `MediaPlayer` lifecycle operations in background jobs without blocking the main UI thread.
   - Confined `audioLock` (`ReentrantLock`) strictly to fast in-memory state mutations so UI thread calls (`pause()`, `play()`, `seekTo()`) are never blocked waiting on I/O.
   - Tracked `activeLoadJob` to cancel stale in-flight loads upon subsequent operations.
3. **Tests**:
   - Synchronized Robolectric `ShadowMediaPlayer` events with coroutines using `testCoroutineDispatchers.runCurrent()`.
   - Added concurrency tests in `AudioPlayerControllerTest` for rapid sequential loads and mid-load teardowns.
   - Added test in `AudioFragmentTest` to verify that repeated state updates with the same audio source keep the audio playing without reloading.

## Essential Checklist
- [x] The PR title starts with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The explanation section above starts with "Fixes #bugnum: " (If this PR fixes part of an issue, use instead: "Fixes part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## Disclosure of LLM Usage
- **Did you use AI/LLMs when working on this PR?** Yes
- **If yes, describe the extent AI was used:** Brainstorming the background coroutine and mutex approach, understanding thread lock contention, and discussing test edge cases.

### CODE REVIEW

**Comment by adhiamboperes** [comment]
@Kishan8548, please be sure to assign a reviewer if the PR is ready.

**Comment by Kishan8548** [comment]
PTAL @Neer-rn

File: app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt [hunk 1]
```diff
@@ -180,8 +181,13 @@ class AudioViewModel @Inject constructor(
     }
 
     if (languageCodeForDataSource != null) {
+      val targetUri = voiceOverToUri(voiceoverMap[languageCodeForDataSource])
+      if (targetUri == currentLoadedAudioUri && reloadingMainContent != true) {
+        return
+      }
+      currentLoadedAudioUri = targetUri
```
[the remark this answers is not in the document]
Comment (Neer-rn) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:188]: Please address this comment.

Comment (Neer-rn) [inline app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:188]: If this audio is already prepared and suppose we call `loadMainContentAudio(allowAutoPlay = true, reloadingContent = false)`, we return here without starting playback, right? Autoplay currently happens when we receive `PREPARED`, but we will not receive another prepared callback when we skip the load. Can we please handle the autoplay request when reusing the same audio? Also, please add a test for this case.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt [hunk 2]
```diff
@@ -115,26 +128,58 @@ class AudioPlayerController @Inject constructor(
...
       playProgress.value =
         AsyncResult.Success(PlayProgress(PlayStatus.COMPLETED, 0, duration))
     }
     mediaPlayer.setOnPreparedListener {
-      prepared = true
-      duration = it.duration
+      audioLock.withLock {
+        prepared = true
+        duration = it.duration
+      }
```
Comment (Neer-rn) [inline domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:182]: Since release now happens in the background, a prepared callback can still arrive after we have marked the player as closed, right? Can we please check that the callback belongs to the current active player and load before updating the state? Also, please add a test for this.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt [hunk 3]
```diff
@@ -277,15 +332,29 @@ class AudioPlayerController @Inject constructor(
...
-      nextUpdateJob?.cancel()
-      nextUpdateJob = null
-      mediaPlayer.reset()
+      stopUpdatingSeekBar()
+      player = mediaPlayer
     }
     playProgress.value =
       AsyncResult.Failure(AudioPlayerException("Audio load aborted before preparation"))
+    activeLoadJob?.cancel()
+    CoroutineScope(backgroundDispatcher).launch {
```
Comment (Neer-rn) [inline domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:345]: What happens if we request another audio load before this reset job runs? This job is not tracked or cancelled, and both requests can use the same player. If the new load runs first, this older job could reset the newly loaded audio. Can we please make sure an old abort cannot reset a newer load? Also, let's add a test.

**Review by Neer-rn — CHANGES_REQUESTED** [review]
Thanks for the work @Kishan8548. The PR mostly looks good, I have left few comments PTAL.

File: app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kishan8548) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:205]: Addressed, resetting `currentLoadedAudioUri` on failure now.

Comment (Kishan8548) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:205]: done, added test for it

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt (hunk 2 above)
Comment (Kishan8548) [reply domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:210]: Done.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt (hunk 3 above)
Comment (Kishan8548) [reply domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt]: Done.

**Comment by Kishan8548** [comment]
@Neer-rn PTAL, addressed all requested changes and added corresponding tests.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt [hunk 4]
```diff
@@ -115,50 +132,113 @@ class AudioPlayerController @Inject constructor(
...
-      mediaPlayer.reset()
-      prepareDataSource(url)
+      player = mediaPlayer
+      currentLoadId++
+      activeLoadId = currentLoadId
+      loadId = currentLoadId
+    }
+    playProgress.value = AsyncResult.Pending()
+    activeLoadJob?.cancel()
+    activeLoadJob = CoroutineScope(backgroundDispatcher).launch {
```
Comment (Neer-rn) [inline domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:150]: Android says that `MediaPlayer` is not thread-safe and all its operations should run on the same thread. Here some operations run on the main thread while others run on `BackgroundDispatcher`, which uses multiple threads. A mutex does not ensure the same thread is used. Can we please use one dedicated background thread with a Looper for all `MediaPlayer` operations?

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt [hunk 5]
```diff
@@ -115,50 +132,113 @@ class AudioPlayerController @Inject constructor(
...
-    mediaPlayer.setOnPreparedListener {
-      prepared = true
-      duration = it.duration
-      playProgress.value =
-        AsyncResult.Success(PlayProgress(PlayStatus.PREPARED, 0, duration))
+    mediaPlayer.setOnPreparedListener { player ->
+      var isCurrentLoad = false
+      audioLock.withLock {
+        if (!isReleased && mediaPlayerActive && player == mediaPlayer &&
+          currentLoadId == activeLoadId && activeLoadId != 0L
```
Comment (Neer-rn) [inline domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:200]: I think a late callback from the old audio can still pass this check because `currentLoadId` and `activeLoadId` are both updated for the new load. I tested this case locally and the old callback changed the status to `PREPARED`. Can we please fix this and add a test where the first audio’s callback arrives after the second audio is requested?

File: app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Neer-rn) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:205]: Thanks, this handles the `Failure` case, but `prepareDataSource()` only logs an `IOException` and does not send a failure. So the URI will still not be cleared in that case. Could you please handle the `IOException` case too and add a test for it?

**Review by Neer-rn — CHANGES_REQUESTED** [review]
@Kishan8548 I again reviewed your PR but with the help of LLM this time and found a few more issues that need to be addressed. Left few comments PTAL.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt (hunk 4 above)
Comment (Kishan8548) [reply domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt]: switched background operations to `@BlockingDispatcher` (Oppia's dedicated single-thread executor), ensuring all background `MediaPlayer` operations run on the exact same thread while preserving deterministic coordination with`TestCoroutineDispatchers`.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt (hunk 5 above)
Comment (Kishan8548) [reply domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt]: addressed, bound `loadId` to each load's listener so late callbacks from an older load don't match `activeLoadId` and are safely ignored. added `testController_secondAudioRequested_firstAudioPreparedCallback_doesNotUpdateState` to verify.

File: app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kishan8548) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt]: addressed, posting failure on `IOException` now so `currentLoadedAudioUri` is cleared. added `testController_prepareDataSource_throwsIoException_capturesFailureState` to test this.

[the remark this answers is not in the document]
Comment (Kishan8548) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt]: Update on this: after testing on CI, emitting `Failure` on `IOException` in `prepareDataSource()` actually breaks `ExplorationActivityTest` (e.g. `testExplorationActivity_loadingAudio_progressbarIsDisplayed`). 

In Oppia's audio flow, a network `IOException` intentionally leaves playback in `Pending()` so that the player stays in `UiAudioPlayStatus.LOADING` and keeps the loading progress indicator visible while waiting for network. Moving to `Failure` changes the status to `FAILED` and hides the indicator (`GONE`), which fails those tests.

Therefore, `prepareDataSource()` preserves `Pending()` on `IOException` as intended, while `Failure` remains for actual playback error events (`OnErrorListener`) and `abortPendingLoad()`.

**Comment by Kishan8548** [comment]
@Neer-rn PTAL, addressed requested changes

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt (hunk 4 above)
Comment (Neer-rn) [reply domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt]: Thanks for switching to `BlockingDispatcher`. `MediaPlayer` is still created and used for play/pause on the UI thread, while reset and preparation run on another thread. Could you make all player access happen on the same thread, as Android requires?

File: app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Neer-rn) [reply app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt]: Keeping `LOADING` makes sense. But for feedback audio, `reloadingMainContent` is false. If `setDataSource()` throws, what retries that same URI? Could you add a test for a failed first attempt followed by a successful request?

**Review by Neer-rn — CHANGES_REQUESTED** [review]
I have left few more comments @Kishan8548.
